Update template element description and add shadowrootmode attribute (#31490)
mdn/content: files/en-us/web/html/element/template/index.md

{{HTMLSidebar}}

> The **`<template>`** [HTML](/en-US/docs/Web/HTML) element is a mechanism for holding {{Glossary("HTML")}} fragments for two main use cases:

Reviewer: Drive by comment: I was reviewing #31577, so was looking for a good `shadowrootmode` explanation, but am not currently reviewing this PR. That said....

The first sentence of each /mdn/content/ page is used as the "summary" by Yari macros and some external sites/software, so I think it may be best to not end in a colon leading into a list as it might get rendered in an unexpected way. See /yari/build/extract-summary.ts for the script that creates the summary.

How about something like:

```suggestion
The **`<template>`** [HTML](/en-US/docs/Web/HTML) element is a mechanism for holding {{Glossary("HTML")}} fragments that can be cloned and inserted in the document by script. The `<template>` element can be used in two ways:
```
In this, I changed "main uses" to "used in two ways" because I think the first is the main use case, but maybe the latter is a main case too. I have not followed this discussion so may be wrong. For now, we can't say "main" though because its not yet fully supported, so, it may become main, but isn't there yet.

Here's a rough draft of an alternative intro (all the way up to attributes): The last paragraph is full of links, but i think this is the gist of the purpose of the template element and needs to be said somehow....

```md
The **`<template>`** [HTML](/en-US/docs/Web/HTML) element is a mechanism for holding {{Glossary("HTML")}} to be instantiated during runtime using JavaScript, rather than rendered immediately when a page is loaded, or to generate a Shadow DOM, or both.

By default, the element's contents are not rendered, only parsed, into a [document fragment](/en-US/docs/Web/API/DocumentFragment). If the `shadowrootmode` attribute is present and set to `open`, the element is replaced in the DOM by its content wrapped in a [shadow root](/en-US/docs/Glossary/Shadow_tree). In this case, the contents of the `<template>` element are rendered in place.

The `<template>` element can be used to [create custom elements](link_to_tutorial_or_example_below) with the elements' contents being the snippet of HTML markup that make up the custom element, with the contents being [cloned](/en-US/docs/Web/API/Node/cloneNode) and inserted into the DOM with JavaScript. Generally, with the [content](/en-US/docs/Web/API/HTMLTemplateElement/content) [appended](/en-US/docs/Web/API/Node/appendChild) to each custom element [attaching a shadow DOM](/en-US/docs/Web/API/Element/attachShadow).
```

Author: @estelle Thank you for the idea. I tried to self-contain the intro and changed use-cases wording to fit. I don’t think we should shy away from describing declarative shadow DOM’s use case on MDN: it will be shipped in Firefox 123 (enabled by default) and already available in Chrome and Safari.

    - `closed`

      - : Hides the internal shadow root DOM from JavaScript.

    > **Note:** This is a feature of the HTML parser that cannot be used after the parsing by setting the `shadowrootmode` attribute via JavaScript. Only allowed values will create the shadow root. Empty or any other value won't trigger this behavior.
    >
    > You may find the non-standard `shadowroot` attribute in older tutorials and examples that used to be supported in Chrome 90-110. This attribute has been removed and replaced by the standard `shadowrootmode` attribute.

The corresponding {{domxref("HTMLTemplateElement")}} interface includes a standard {{domxref("HTMLTemplateElement.content", "content")}} property (without an equivalent content/markup attribute). This `content` property is read-only and holds a {{domxref("DocumentFragment")}} that contains the DOM subtree represented by the template. Be careful when using the `content` property because the returned `DocumentFragment` can exhibit unexpected behavior. For more details, see the [Avoiding DocumentFragment pitfalls](#avoiding_documentfragment_pitfalls) section below.

## Examples

> ### Table row generation

Reviewer 2: Thanks for adding a title for this example! I'd suggest user action-based titles:
```suggestion
### Generating table rows
```

{{HTMLSidebar}}

The **`<template>`** [HTML](/en-US/docs/Web/HTML) element serves as a mechanism for holding {{Glossary("HTML")}} fragments, which can either be used later via JavaScript or generated immediately into shadow DOM.

> 1. By default, the element's content is not rendered, only parsed into a [document fragment](/en-US/docs/Web/API/DocumentFragment). Using the {{domxref("HTMLTemplateElement.content", "content")}} property in JavaScript, this fragment can be cloned via the {{domxref("Node.cloneNode", "cloneNode")}} method and inserted into the DOM.
> 2. If the element contains the `shadowrootmode` attribute, the HTML parser will immediately generate a shadow DOM. The element is replaced in the DOM by its content wrapped in a [shadow root](/en-US/docs/Glossary/Shadow_tree).

Reviewer 2: Reconsidering these two points.
- Can we move this content to the "Usage notes" section (before the "The corresponding `HTMLTemplateElement` interface" paragraph)? This is basically a more detailed explanation of the one-liner introductory para.
- Also, the first sentence describes the default state and then there are two possibilities. Can we reword this to be explicit about what users can do? What do you think about:
```suggestion
By default, an element's content is not rendered, only parsed into a [document fragment](/en-US/docs/Web/API/DocumentFragment). You can use this fragment in the following ways:
- If you use the {{domxref("HTMLTemplateElement.content", "content")}} property in JavaScript, you can clone this fragment via the {{domxref("Node.cloneNode", "cloneNode")}} method and insert it into the DOM.
- If you add the `shadowrootmode` attribute to the element, the HTML parser will immediately generate a shadow DOM. The element is replaced in the DOM by its content wrapped in a [shadow root](/en-US/docs/Glossary/Shadow_tree).
```

Reviewer 2: This content is now repeated twice, once at the top and second time in "Usage notes". Did you want to keep it only at the top of the page?

Author: Oh, I guess I merged something wrong. Now it’s only in usage notes. Thanks!

## Attributes

This element includes the [global attributes](/en-US/docs/Web/HTML/Global_attributes).

- `shadowrootmode`

  - : Creates a [shadow root](/en-US/docs/Glossary/Shadow_tree) for the parent element. It is a declarative version of the {{domxref("Element.attachShadow()")}} method and accepts the same values.

    - `open`

      - : Exposes the internal shadow root DOM for JavaScript (recommended for most use cases).

    - `closed`

      - : Hides the internal shadow root DOM from JavaScript.

    > **Note:** This is a feature of the HTML parser that cannot be used post-parsing by setting the `shadowrootmode` attribute through JavaScript. Only allowed values will create the shadow root; any other values, including empty ones, won't trigger this behavior.
    >

>     > You may find the non-standard `shadowroot` attribute in older tutorials and examples that used to be supported in Chrome 90-110. This attribute has since been removed and replaced by the standard `shadowrootmode` attribute.

Reviewer 2: I was wondering if we could link to the documentation for `shadowroot` attribute  because the BCD table could indicate the browser support. But we don't have this attribute on MDN. In that case, maybe we shouldn't add this note here specifically for Chrome.

Author: This is something worth mentioning because the previous Chrome-only implementation of the declarative shadow DOM has been around for two years with a polyfill as an option for non-supporting browsers. Now that all browsers have this attribute's new version, there might still be the old version in some projects’ code. I want this information to be available here.

By the way, it’s actually [mentioned](https://github.com/mdn/browser-compat-data/blob/main/html/elements/template.json#L49) in BCD.

Reviewer 2: > This is something worth mentioning because the previous Chrome-only implementation of the declarative shadow DOM has been around for two years with a polyfill as an option for non-supporting browsers.

👍 This is helpful, thanks! Agree to keep this note with this context

> By the way, it’s actually [mentioned](https://github.com/mdn/browser-compat-data/blob/main/html/elements/template.json#L49) in BCD.

Good one! More support for keeping that note then!

Reviewer 2: What do you think about creating two separate note boxes though? The first one is about parsing and the second one is a completely different topic, about `shadowroot`.

Author: Good point! Done :)

Reviewer 2: Thanks!

{{HTMLSidebar}}

The **`<template>`** [HTML](/en-US/docs/Web/HTML) element serves as a mechanism for holding {{Glossary("HTML")}} fragments, which can either be used later via JavaScript or generated immediately into shadow DOM.

By default, an element's content is not rendered, only parsed into a [document fragment](/en-US/docs/Web/API/DocumentFragment). You can use this fragment in the following ways:

- If you use the {{domxref("HTMLTemplateElement.content", "content")}} property in JavaScript, you can clone this fragment via the {{domxref("Node.cloneNode", "cloneNode")}} method and insert it into the DOM.
- If you add the `shadowrootmode` attribute to the element, the HTML parser will immediately generate a shadow DOM. The element is replaced in the DOM by its content wrapped in a [shadow root](/en-US/docs/Glossary/Shadow_tree).

## Attributes

This element includes the [global attributes](/en-US/docs/Web/HTML/Global_attributes).

- `shadowrootmode`

>   - : Creates a [shadow root](/en-US/docs/Glossary/Shadow_tree) for the parent element. It is a declarative version of the {{domxref("Element.attachShadow()")}} method and accepts the same values.

Reviewer: ```suggestion
  - : Creates a [shadow root](/en-US/docs/Glossary/Shadow_tree) for the parent element. It is a declarative version of the {{domxref("Element.attachShadow()")}} method and accepts the same {{glossary("enumerated")}} values.
```

}
```

{{EmbedLiveSample("Table row generation", 500, 120)}}

> ### Declarative shadow DOM

Reviewer 2: what about:
```suggestion
### Implementing a declarative shadow DOM
```

> At the beginning of the markup, we have a hidden support warning that we will show later via JavaScript if the browser doesn't support the `shadowrootmode` attribute. After that, there are two {{HTMLElement("article")}} elements with nested {{HTMLElement("style")}} elements that behave differently. The first is global to the whole document, while the second is scoped to the shadow root generated in place of the `<template>` element because the `shadowrootmode` attribute is present.

Reviewer 2: Just a style preference to make the sentences more compact
```suggestion
In this example, a hidden support warning is included at the beginning of the markup. This warning is later set to be displayed via JavaScript if the browser doesn't support the `shadowrootmode` attribute. Next, there are two {{HTMLElement("article")}} elements, each containing nested {{HTMLElement("style")}} elements with different behaviors. The first `<style>` element is global to the whole document. The second one is scoped to the shadow root generated in place of the `<template>` element because of the presence of the `shadowrootmode` attribute.
```

```html
<p hidden>
  ⛔ Your browser doesn't support <code>shadowrootmode</code> attribute yet.
</p>
<article>
  <style>
    p {
      padding: 8px;
      background-color: wheat;
    }
  </style>
  <p>I'm in the DOM.</p>
</article>
<article>
  <template shadowrootmode="open">
    <style>
      p {
        padding: 8px;
        background-color: plum;
      }
    </style>

>     <p>I'm in the Shadow DOM.</p>

Reviewer 2: ```suggestion
    <p>I'm in the shadow DOM.</p>
```

- [`ShadowRoot`]("/en-US/docs/Web/API/ShadowRoot) interface
- [Using templates and slots](/en-US/docs/Web/API/Web_components/Using_templates_and_slots)
- [CSS scoping](/en-US/docs/Web/CSS/CSS_scoping) module

> - [Introduction to declarative shadow DOM](https://developer.chrome.com/docs/css-ui/declarative-shadow-dom)

Reviewer 2: We have a [few guidelines](https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines/Writing_style_guide#see_also_section) for when we add external links to "See also", namely specifying the source site, mentioning the last update date and matching the title of the external page but still following sentence casing on MDN
```suggestion
- [Declarative shadow DOM](https://developer.chrome.com/docs/css-ui/declarative-shadow-dom) on developer.chrome.com (2023)
```
